hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/ITestRestrictedReadAccess.java

```diff
@@ -0,0 +1,459 @@
+        ContractTestUtils.readUTF8(roleFS, subdirFile, data.length));
+    if (authMode) {
+      // auth mode doesn't just not check the store,
+      // because it knows the file length is zero,
+      // it returns -1 without even opening the file.
+      try (FSDataInputStream is = roleFS.open(emptyFile)) {
+        Assertions.assertThat(is.read())
+            .describedAs("read of empty file")
+            .isEqualTo(-1);
+      }
+      roleFS.getFileStatus(subdirFile);
+    } else {
+      // non-auth mode, it fails at some point in the opening process.
+      intercept(AccessDeniedException.class, () ->
+          ContractTestUtils.readUTF8(roleFS, emptyFile, 0));
+    }
+
+    // Fun with Glob
+    describe("Glob Status operations");
+    // baseline: the real filesystem on a subdir
+    globFS(realFS, subdirFile, null, false, 1);
+    // a file fails if not in auth mode
+    globFS(roleFS, subdirFile, null, !authMode, 1);
+    // empty directories don't fail.
+    assertStatusPathEquals(emptyDir,
+        globFS(roleFS, emptyDir, null, false, 1));
+
+    Path textFilePathPattern = new Path(noReadDir, "*/*.txt");
+    FileStatus[] st = globFS(roleFS,
+        textFilePathPattern,
+        null, false, 2);
+    Assertions.assertThat(st)
+        .extracting("getPath")
+        .containsExactlyInAnyOrder(subdirFile, subdir2File1);
+
+    globFS(roleFS,
+        new Path(noReadDir, "*/*.docx"),
+        null, false, 1);
+    globFS(roleFS,
+        new Path(noReadDir, "*/*.doc"),
+        null, false, 0);
+    globFS(roleFS, noReadDir,
+        ANYTHING, false, 1);
+
+    // now run a located file status fetcher against it
+    describe("LocatedFileStatusFetcher operations");
+
+    // use the same filter as FileInputFormat; single thread.
+    roleConfig.setInt(LIST_STATUS_NUM_THREADS, 1);
+    LocatedFileStatusFetcher fetcher =
+        new LocatedFileStatusFetcher(
+            roleConfig,
+            new Path[]{basePath},
+            true,
+            HIDDEN_FILE_FILTER,
+            true);
+    Assertions.assertThat(fetcher.getFileStatuses())
+        .describedAs("result of located scan")
+        .flatExtracting(FileStatus::getPath)
+        .containsExactlyInAnyOrder(
+            emptyFile,
+            subdirFile,
+            subdir2File1,
+            subdir2File2);
+
+    describe("LocatedFileStatusFetcher with %s", textFilePathPattern);
+    roleConfig.setInt(LIST_STATUS_NUM_THREADS, 4);
+    LocatedFileStatusFetcher fetcher2 =
+        new LocatedFileStatusFetcher(
+            roleConfig,
+            new Path[]{textFilePathPattern},
+            true,
+            ANYTHING,
+            true);
+    Assertions.assertThat(fetcher2.getFileStatuses())
+        .describedAs("result of located scan")
+        .isNotNull()
+        .flatExtracting(FileStatus::getPath)
+        .containsExactlyInAnyOrder(subdirFile, subdir2File1);
+
+    // query a file and expect the initial scan to fail except in
```

Comment:
it's ok to group these, but we could at least have factored out private methods and call it from here

Reply:
split up